contracts/javascore/xcall-multi-protocol/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java

```diff
@@ -0,0 +1,352 @@
+    public void CallExecuted(BigInteger _reqId, int _code, String _msg) {}
+
+    @Override
+    @EventLog(indexed=1)
+    public void ResponseMessage(BigInteger _sn, int _code, String _msg) {}
+
+    @Override
+    @EventLog(indexed=1)
+    public void RollbackMessage(BigInteger _sn) {}
+
+    @Override
+    @EventLog(indexed=1)
+    public void RollbackExecuted(BigInteger _sn, int _code, String _msg) {}
+
+    @Override
+    @EventLog(indexed=3)
+    public void CallMessageSent(Address _from, String _to, BigInteger _sn) {}
+
+    /* ========== Interfaces with BMC ========== */
+    @External
+    public void handleBTPMessage(String _from, String _svc, BigInteger _sn, byte[] _msg) {
+        CSMessage msg = CSMessage.fromBytes(_msg);
+        switch (msg.getType()) {
+            case CSMessage.REQUEST:
+                handleRequest(_from, _sn, msg.getData());
+                break;
+            case CSMessage.RESPONSE:
+                handleResponse(_from, _sn, msg.getData());
+                break;
+            default:
+                Context.revert("UnknownMsgType(" + msg.getType() + ")");
+        }
+    }
+
+    @External
+    public void handleBTPError(String _src, String _svc, BigInteger _sn, long _code, String _msg) {
+        String errMsg = "BTPError{code=" + _code + ", msg=" + _msg + "}";
+        CSMessageResponse res = new CSMessageResponse(_sn, CSMessageResponse.BTP_ERROR, errMsg);
+        handleResponse(_src, _sn, res.toBytes());
+    }
+    /* ========================================= */
+
+    private BigInteger sendBTPMessage(Address _connection, BigInteger value, String netTo, int msgType, BigInteger sn, byte[] data) {
+        CSMessage msg = new CSMessage(msgType, data);
+        ConnectionScoreInterface connection = new ConnectionScoreInterface(_connection);
+        return connection.sendMessage(value, netTo, NAME, sn, msg.toBytes());
+    }
+
+    private void handleRequest(String netFrom, BigInteger sn, byte[] data) {
+        CSMessageRequest msgReq = CSMessageRequest.fromBytes(data);
+        String[] protocols = msgReq.getProtocols();
+        String from = msgReq.getFrom();
+        Context.require(NetworkAddress.valueOf(from).net().equals(netFrom));
+
+        if (protocols.length > 1) {
+            byte[] hash = Context.hash("sha-256", data);
+            String source = Context.getCaller().toString();
+            pendingReqs.at(hash).set(source, true);
+            for (String protocol : protocols) {
+                if (!pendingReqs.at(hash).getOrDefault(protocol, false)) {
+                    return;
+                }
+            }
+
+            for (String protocol : protocols) {
+                pendingReqs.at(hash).set(protocol, null);
+            }
+        }
+
+        String to = msgReq.getTo();
+        BigInteger reqId = getNextReqId();
+        CSMessageRequest req = new CSMessageRequest(msgReq.getFrom(), to, msgReq.getProtocols(), msgReq.getSn(), msgReq.needRollback(), msgReq.getData());
```

Comment:
Is there a difference between req and msgReq message requests?

Reply:
No not currently :+1: